File: engine/can_engine/framing.py
"""Length-prefixed JSON frames: 4-byte big-endian size + UTF-8 payload."""

from __future__ import annotations

import json
import struct
from typing import Any, Final

MAX_FRAME_BYTES: Final[int] = 1_048_576
_HEADER = struct.Struct(">I")


class ProtocolError(Exception):
    """Malformed length or JSON. The server must not crash on this."""

    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code
        self.message = message


def encode_message(message: dict[str, Any]) -> bytes:
    payload = json.dumps(message, separators=(",", ":"), ensure_ascii=False).encode("utf-8")
    if len(payload) > MAX_FRAME_BYTES:
        raise ProtocolError("payload_too_large", "encoded JSON exceeds 1 MiB")
    return _HEADER.pack(len(payload)) + payload


def decode_payload(payload: bytes) -> dict[str, Any]:
    if not payload:
        raise ProtocolError("invalid_length", "empty JSON payload")
    try:
        parsed = json.loads(payload.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise ProtocolError("invalid_json", f"JSON decode failed: {exc}") from exc
    if not isinstance(parsed, dict):
        raise ProtocolError("invalid_message", "JSON root must be an object")
    msg_type = parsed.get("type")
    if not isinstance(msg_type, str) or not msg_type:
        raise ProtocolError("invalid_message", "message type must be a non-empty string")
    return parsed


def check_length(length: int) -> None:
    if length <= 0:
        raise ProtocolError("invalid_length", f"length must be > 0, got {length}")
    if length > MAX_FRAME_BYTES:
        raise ProtocolError("payload_too_large", f"length {length} exceeds 1 MiB")

File: electron/main/framing.ts
/** 4-byte big-endian length + UTF-8 JSON (same as engine/can_engine/framing.py). */

export const MAX_FRAME_BYTES = 1_048_576

export class ProtocolError extends Error {
  readonly code: string

  constructor(code: string, message: string) {
    super(message)
    this.name = 'ProtocolError'
    this.code = code
  }
}

export function encodeMessage(message: unknown): Buffer {
  const payload = Buffer.from(JSON.stringify(message), 'utf8')
  if (payload.length > MAX_FRAME_BYTES) {
    throw new ProtocolError('payload_too_large', 'encoded JSON exceeds 1 MiB')
  }
  const header = Buffer.allocUnsafe(4)
  header.writeUInt32BE(payload.length, 0)
  return Buffer.concat([header, payload])
}

export class FrameDecoder {
  private buffer = Buffer.alloc(0)

  push(chunk: Buffer): Record<string, unknown>[] {
    this.buffer = Buffer.concat([this.buffer, chunk])
    const messages: Record<string, unknown>[] = []

    while (this.buffer.length >= 4) {
      const length = this.buffer.readUInt32BE(0)
      if (length <= 0) {
        throw new ProtocolError('invalid_length', `length must be > 0, got ${length}`)
      }
      if (length > MAX_FRAME_BYTES) {
        throw new ProtocolError('payload_too_large', `length ${length} exceeds 1 MiB`)
      }
      if (this.buffer.length < 4 + length) {
        break
      }
      const json = this.buffer.subarray(4, 4 + length).toString('utf8')
      this.buffer = this.buffer.subarray(4 + length)
      let parsed: unknown
      try {
        parsed = JSON.parse(json)
      } catch (error) {
        const detail = error instanceof Error ? error.message : String(error)
        throw new ProtocolError('invalid_json', `JSON decode failed: ${detail}`)
      }
      if (parsed === null || typeof parsed !== 'object' || Array.isArray(parsed)) {
        throw new ProtocolError('invalid_message', 'JSON root must be an object')
      }
      const record = parsed as Record<string, unknown>
      if (typeof record.type !== 'string' || record.type.length === 0) {
        throw new ProtocolError('invalid_message', 'message type must be a non-empty string')
      }
      messages.push(record)
    }

    return messages
  }

  reset(): void {
    this.buffer = Buffer.alloc(0)
  }
}
